From aac36163dcbf1df343b1c2776e042433b656b4ec Mon Sep 17 00:00:00 2001
Message-Id: <aac36163dcbf1df343b1c2776e042433b656b4ec.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Michel Thierry <michel.thierry@intel.com>
Date: Mon, 7 Jul 2014 14:38:16 +0100
Subject: [PATCH 139/312] FOR_UPSTREAM [VPG]: drm/i915: Add
 WaIncreaseTagClockTimer:bdw

Workaround requires programing L3 tag clock timing register.

Issue: VIZ-4086
Change-Id: I443b94f3faf997f93d1651b2feb54c91db93d2ca
Signed-off-by: Michel Thierry <michel.thierry@intel.com>
---
 drivers/gpu/drm/i915/i915_reg.h |    4 ++++
 drivers/gpu/drm/i915/intel_pm.c |    9 +++++++++
 2 files changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 97f70a1..9075185 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -4744,6 +4744,10 @@ enum punit_power_well {
 #define GEN8_L3SQCREG1				0xb100
 #define  GEN8_L3SQCREG1_DEFAULT_VALUE	0x784000
 
+#define GEN8_L3CNTLREG1				0xb10c
+#define  GEN8_TAG_CLK_OFFTIME			(1<<23)
+#define  GEN8_TAG_CLK_OFFTIME_MASK		(~((1<<23) | (1<<22) | (1<<21) | (1<<20)))
+
 #define GEN8_L3SQCREG4				0xb118
 #define  GEN8_PIPELINE_FLUSH_COHERENT_LINES	(1<<21)
 
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index 6dc3b37..48f6882 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -5618,6 +5618,15 @@ static void gen8_init_clock_gating(struct drm_device *dev)
 	/* WaFlushCoherentL3CacheLinesAtContextSwitch:bdw */
 	I915_WRITE(GEN8_L3SQCREG4, I915_READ(GEN8_L3SQCREG4) |
 			GEN8_PIPELINE_FLUSH_COHERENT_LINES);
+
+	/*
+	 * WaIncreaseTagClockTimer:bdw
+	 * This register is not masked and the default value is not all zero.
+	 * Use the mask with and to clear out bits 23:20.
+	 */
+	I915_WRITE(GEN8_L3CNTLREG1, (I915_READ(GEN8_L3CNTLREG1) &
+			GEN8_TAG_CLK_OFFTIME_MASK) |
+			GEN8_TAG_CLK_OFFTIME);
 }
 
 static void haswell_init_clock_gating(struct drm_device *dev)
-- 
1.7.9.5

